mono/metadata/assembly.c

```diff
@@ -1460,14 +1468,18 @@ load_reference_by_aname_refonly_asmctx (MonoAssemblyName *aname, MonoAssembly *a
 			MonoAssemblyByNameRequest req;
 			memset (&req, 0, sizeof (req));
 			req.request.asmctx = MONO_ASMCTX_DEFAULT;
+			req.request.want_pinned = TRUE;
 			req.requesting_assembly = assm;
 			req.basedir = assm->basedir;
 			reference = mono_assembly_request_byname (aname, &req, status);
 		} else {
 			reference = mono_assembly_loaded_full (aname, TRUE);
-			if (!reference)
+			mono_assembly_pinref (reference);
+			if (!reference) {
 				/* Try a postload search hook */
 				reference = mono_assembly_invoke_search_hook_internal (aname, assm, TRUE, TRUE);
+				mono_assembly_pinref (reference);
```

Comment:
Won't this introduce a race condition as well? We have to increment while holding the lock coordinating unload.